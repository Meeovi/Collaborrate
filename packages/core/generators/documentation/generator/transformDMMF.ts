import { DMMF as ExternalDMMF } from '@prisma/generator-helper';

export function lowerCase(name: string): string {
  return name.substring(0, 1).toLowerCase() + name.substring(1);
}

export interface DMMFMapping {
  model: string;
  findOne?: string | null;
  findFirst?: string | null;
  findMany?: string | null;
  create?: string | null;
  update?: string | null;
  updateMany?: string | null;
  upsert?: string | null;
  delete?: string | null;
  deleteMany?: string | null;
}

export type DMMFDocument = Omit<ExternalDMMF.Document, 'mappings'> & {
  mappings: DMMFMapping[];
};

export default function transformDMMF(
  dmmf: ExternalDMMF.Document
): DMMFDocument {
  return {
    ...dmmf,
    mappings: getMappings(dmmf.mappings, dmmf.datamodel),
  };
}

function getMappings(
  mappings: ExternalDMMF.Mappings,
  datamodel: ExternalDMMF.Datamodel
): DMMFMapping[] {
  const modelOperations = mappings.modelOperations
    .filter((mapping) => {
      const model = datamodel.models.find((m) => m.name === mapping.model);
      if (!model) {
        throw new Error(`Mapping without model ${mapping.model}`);
      }
      return model.fields.some((f) => f.kind !== 'object');
    })
    .map((mapping: any) => ({
      model: mapping.model,
      findUnique: mapping.findSingle || mapping.findOne || mapping.findUnique,
      findFirst: mapping.findFirst,
      findMany: mapping.findMany,
      create: mapping.createOne || mapping.createSingle || mapping.create,
      delete: mapping.deleteOne || mapping.deleteSingle || mapping.delete,
      update: mapping.updateOne || mapping.updateSingle || mapping.update,
      deleteMany: mapping.deleteMany,
      updateMany: mapping.updateMany,
      upsert: mapping.upsertOne || mapping.upsertSingle || mapping.upsert,
    }));

  return modelOperations;
}
