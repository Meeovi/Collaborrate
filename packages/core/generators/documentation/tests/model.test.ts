import ModelGenerator from '../generator/model';
import transformDMMF from '../generator/transformDMMF';
//@ts-ignore
import { getDMMF } from '@prisma/sdk';

describe('model generator', () => {
  it('renders model directive html correctly', async () => {
    const datamodelString = /* Prisma */ `
      model User {
        id String @default(cuid())
        name String
        @@index([name])
        @@id([name, id])
        @@unique([name])
      }
    `;

    const dmmf = await getDMMF({ datamodel: datamodelString });
    const transformedDmmf = transformDMMF(dmmf);

    const modelGen = new ModelGenerator(transformedDmmf);
    const spy = jest.spyOn(modelGen, 'getModelDiretiveHTML');

    // Run the generator against the model
    modelGen.toHTML();

    expect(spy).toHaveBeenCalledTimes(3);
    expect(spy).toHaveBeenCalledWith({
      name: '@@index',
      values: ['name'],
    });
    expect(spy).toHaveBeenCalledWith({
      name: '@@id',
      values: ['name', 'id'],
    });
    expect(spy).toHaveBeenCalledWith({
      name: '@@unique',
      values: ['name'],
    });

    // snapshot for record
    expect(
      modelGen.getModelDiretiveHTML({ name: '@@id', values: ['name', 'id'] })
    ).toMatchSnapshot();
  });

  it('renders model field table row correctly', async () => {
    const datamodelString = /* Prisma */ `
      model User {
        id String @id @default(cuid())
        name String
        /// this the email of the user
        email String @unique
        password String
      }

      model Post {
        id String @id @default(cuid())
        title String?
      }
    `;

    const dmmf = await getDMMF({ datamodel: datamodelString });
    const transformedDmmf = transformDMMF(dmmf);

    const modelGen = new ModelGenerator(transformedDmmf);
    const spy = jest.spyOn(modelGen, 'getModelFieldTableRow');

    // Run the generator against the model
    modelGen.toHTML();

    expect(spy).toHaveBeenCalledTimes(6);
    expect(spy).toHaveBeenCalledWith(
      {
        name: 'id',
        type: 'String',
        bareTypeName: 'String',
        directives: ['@id', '@default(cuid())'],
        required: true,
      },
      'User'
    );

    expect(spy).toHaveBeenCalledWith(
      {
        name: 'email',
        type: 'String',
        bareTypeName: 'String',
        directives: ['@unique'],
        documentation: 'this the email of the user',
        required: true,
      },
      'User'
    );

    expect(spy).toHaveBeenCalledWith(
      {
        name: 'title',
        type: 'String?',
        bareTypeName: 'String',
        directives: [],
        required: false,
      },
      'Post'
    );

    expect(
      modelGen.getModelFieldTableRow(
        {
          name: 'title',
          type: 'String?',
          bareTypeName: 'String',
          directives: [],
          required: false,
        },
        'Post'
      )
    ).toMatchSnapshot();

    expect(
      modelGen.getModelFieldTableRow(
        {
          name: 'email',
          type: 'String',
          bareTypeName: 'String',
          directives: [],
          documentation: 'This is some docs',
          required: true,
        },
        'User'
      )
    ).toMatchSnapshot();
  });

  it('renders model operation html correctly', async () => {
    const datamodelString = /* Prisma */ `
      model User {
        id String @default(cuid()) @id
        name String
        otherField Int
        posts Post[]
      }

      model Post {
        id String @id @default(cuid())
        title String?
        userId String
        user User @relation(fields:[userId], references:[id])
      }
    `;

    const dmmf = await getDMMF({ datamodel: datamodelString });
    const transformedDmmf = transformDMMF(dmmf);

    const modelGen = new ModelGenerator(transformedDmmf);
    const spy = jest.spyOn(modelGen, 'getModelOperationMarkup');
    modelGen.toHTML();

    expect(spy).toHaveBeenCalled();

    expect(
      modelGen.getModelOperationMarkup(
        {
          name: 'findOne',
          description: 'Find zero or one user',
          usage: 'Example usage',
          opKeys: [
            {
              name: 'where',
              required: false,
              types: [
                {
                  type: 'UserWhereInput',
                  isList: false,
                  location: 'inputObjectTypes',
                },
              ],
            },
          ],
          output: { type: 'User', required: true, list: false },
        },
        'User'
      )
    ).toMatchSnapshot();
  });

  // TODO: add more tests for operations transform
});
