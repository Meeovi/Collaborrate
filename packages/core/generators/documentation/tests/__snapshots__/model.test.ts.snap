// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model generator renders model directive html correctly 1`] = `
"
      <tr>
        <td class=\\"px-4 py-2 border\\">
         <strong>@@id</strong>
        </td>

        <td class=\\"px-4 py-2 border\\"> <ul>
            <li>name</li><li>id</li>
          </ul>
        </td>
      </tr>
    "
`;

exports[`model generator renders model field table row correctly 1`] = `
"
    <tr id=\\"model-Post-title\\">
      <td class=\\"px-4 py-2 border\\">
       title 
      </td>
      <td class=\\"px-4 py-2 border\\">
       String?
      </td>
      <td class=\\"px-4 py-2 border\\">
        <ul>
          <li> - </li>
        </ul>
      </td>
      <td class=\\"px-4 py-2 border\\">
        No
      </td>
      <td class=\\"px-4 py-2 border\\">
        -
      </td>
    </tr>
    "
`;

exports[`model generator renders model field table row correctly 2`] = `
"
    <tr id=\\"model-User-email\\">
      <td class=\\"px-4 py-2 border\\">
       email 
      </td>
      <td class=\\"px-4 py-2 border\\">
       String
      </td>
      <td class=\\"px-4 py-2 border\\">
        <ul>
          <li> - </li>
        </ul>
      </td>
      <td class=\\"px-4 py-2 border\\">
        <b>Yes</b>
      </td>
      <td class=\\"px-4 py-2 border\\">
        This is some docs
      </td>
    </tr>
    "
`;

exports[`model generator renders model operation html correctly 1`] = `
"
                <div class=\\"mt-4\\">
                  <h4 id=\\"model-User-findOne\\" class=\\"mb-2 text-lg font-bold\\">findOne</h4>
                  <p>Find zero or one user</p>
                  <div class=\\"mb-2\\">
                    <pre
                      class=\\"language-markup\\"
                    ><code class=\\" language-javascript\\">Example usage</code></pre>
                  </div>
                  <h4 class=\\"text-lg mb-2\\">Input</h4>
                  <table class=\\"table-auto mb-2\\">
                    <thead>
                      <tr>
                        <th class=\\"px-4 py-2 border\\">Name</th>
                        <th class=\\"px-4 py-2 border\\">Type</th>
                        <th class=\\"px-4 py-2 border\\">Required</th>
                      </tr>
                    </thead>
                    <tbody>
                      
                      <tr>
                        <td class=\\"px-4 py-2 border\\">
                          where
                        </td>
                        <td class=\\"px-4 py-2 border\\">
                        <a href=\\"#type-inputType-UserWhereInput\\">UserWhereInput</a>
                        </td>
                        <td class=\\"px-4 py-2 border\\">
                         No 
                        </td>
                      </tr>
                      
                    </tbody>
                  </table>
                  <h4 class=\\"text-lg mb-2\\">Output</h4>
                  <div><strong>Type: </strong> <a href=\\"#type-outputType-User\\">User</a></div>
                  <div><strong>Required: </strong>
                  Yes</div>
                  <div><strong>List: </strong>
                  No</div>
              </div>
    "
`;
